packages/typespec-vscode/src/npm-utils.ts

```diff
@@ -0,0 +1,143 @@
+import fs from "fs";
+import path from "path";
+import semver from "semver";
+import logger from "./log/logger.js";
+import { ExecOutput, loadModule, spawnExecution, spawnExecutionEvents } from "./utils.js";
+
+export enum InstallationAction {
+  Install = "Install",
+  Upgrade = "Upgrade",
+  Skip = "Skip",
+  Cancel = "Cancel",
+}
+
+export enum npmDependencyType {
+  dependencies = "dependencies",
+  peerDependencies = "peerDependencies",
+  devDependencies = "devDependencies",
+}
+
+export interface NpmPackageInfo {
+  name: string;
+  version?: string;
+  resolved?: string;
+  overridden?: string;
+  dependencies?: Record<string, NpmPackageInfo>;
+}
+
+export class NpmUtil {
+  private cwd: string;
+
+  constructor(cwd: string) {
+    this.cwd = cwd;
+  }
+
+  public async npmInstallPackages(
+    packages: string[] = [],
+    options: any = {},
+    on?: spawnExecutionEvents,
+  ): Promise<ExecOutput> {
+    return spawnExecution("npm", ["install", ...packages], this.cwd, on);
+  }
+
+  public async ensureNpmPackageInstall(
+    packageName: string,
+    version?: string,
+  ): Promise<{ action: InstallationAction; version?: string }> {
+    const { installed: isPackageInstalled, version: installedVersion } =
+      await this.isPackageInstalled(packageName);
+    if (isPackageInstalled) {
+      if (version && installedVersion !== version) {
+        return {
+          action: InstallationAction.Upgrade,
+          version: version,
+        };
+      }
+      return { action: InstallationAction.Skip, version: installedVersion };
+    } else {
+      return { action: InstallationAction.Install, version: version };
+    }
+  }
+
+  public async ensureNpmPackageDependencyInstall(
```

Comment:
seems you are just figure out the missing dependencies, the name is confusing here...

Reply:
This is to figure out the dependency which may has conflict (the package already installed, but it does not match the requisites ), and need to upgrade. So it is to check out which dependency package need to be upgraded.

Reply:
update the name to `ensureNpmPackageDependencyToUpgrade` and comment to make it clear

Reply:
the name is still confusing to me. @timotheeguerin , @chrisradek to see whether they have better idea for it.